presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -81,6 +79,9 @@
     @GuardedBy("exclusiveLock")
     private TaskSource currentTaskSource;
 
+    @GuardedBy("exclusiveLock")
+    private ListenableFuture<?> previouslyReturnedBlockedFuture = NOT_BLOCKED;
```

Comment:
Is this an optimization or essential for the correctness of the implementation? If the former, make this change in a separate commit.